### PR TITLE
342 mtj support

### LINKED ISSUE #342: [EPIC] multi-task jobs - enable support for
multi-task jobs are a fairly new feature to the Databricks product and likely throw some curve balls into the current logic. A review is needed to validate the logic and correct it where necessary to account for multi-task jobs.

### LINKED ISSUE #411: [FEAT] - add jobs custom tags
jobs recently received a feature that enables tagging on the job. job tags should be converted to map throughout the pipelines and surface in gold layer in the job (and perhaps jobrun) table in gold + presentation.

### LINKED ISSUE #445: job tags - struct to map
Databricks recently released a new "job tags" feature. This is currently making its way into job_gold as a struct instead of a map k:v. Convert this field from struct to map string:string just as all other user-defined structs.

Must be part of a minor release as it will require a schema change to all tables with this field in it -- they must be upgraded from tags struct to tags map string:string

### PR DESCRIPTION
**Pull request by GeekSheikh**
closes #342 
closes #411 
closes #445

### CODE REVIEW

File: src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctions.scala [hunk 1]
```diff
@@ -165,6 +165,22 @@ object PipelineFunctions {
...
+   * @param newClusterPrefix path to new_cluster including the "new_cluster" such as "new_settings.new_cluster"
+   * @return Map[String, Column] for changeInventory
+   */
+  def newClusterCleaner(df: DataFrame, newClusterPrefix: String): Map[String, Column] = {
+    Map(
+      s"${newClusterPrefix}.custom_tags" -> SchemaTools.structToMap(df, s"${newClusterPrefix}.custom_tags"),
+      s"${newClusterPrefix}.spark_conf" -> SchemaTools.structToMap(df, s"${newClusterPrefix}.spark_conf"),
+      s"${newClusterPrefix}.spark_env_vars" -> SchemaTools.structToMap(df, s"${newClusterPrefix}.spark_env_vars"),
+      s"${newClusterPrefix}.aws_attributes" -> SchemaTools.structToMap(df, s"${newClusterPrefix}.aws_attributes"),
+      s"${newClusterPrefix}.azure_attributes" -> SchemaTools.structToMap(df, s"${newClusterPrefix}.azure_attributes")
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctions.scala:180]: should we add `gcp_attributes` in advance? Maybe just add comment not to forget it

Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctions.scala:180]: It should stay as struct - it's controlled by us, so problems with changes by user won't happen

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala [hunk 2]
```diff
@@ -258,11 +284,87 @@ object Schema extends SparkSessionWrapper {
...
+      )), nullable = true),
+      StructField("whl",StringType, nullable = true)
+    )), containsNull = true
+  )
+
+  val minimumManualOverrideParamsSchema: StructType = StructType(Seq(
+    StructField("jar_params",ArrayType(StringType, containsNull = true), nullable = true),
+    StructField("python_params",ArrayType(StringType, containsNull = true), nullable = true),
+    StructField("spark_submit_params",ArrayType(StringType, containsNull = true), nullable = true),
+    StructField("notebook_params",ArrayType(StringType, containsNull = true), nullable = true)
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:305]: also add `python_named_params`, `pipeline_params`, `sql_params` and `dbt_commands`: https://docs.databricks.com/dev-tools/api/latest/jobs.html#operation/JobsRunNow

You

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala [hunk 3]
```diff
@@ -258,11 +284,87 @@ object Schema extends SparkSessionWrapper {
...
+
+  val minimumPipelineTaskSchema: StructType = StructType(Seq(StructField("pipeline_id",StringType,nullable = true)))
+
+  val minimumPythonWheelTaskSchema: StructType = StructType(Seq(
+    StructField("entry_point",StringType, nullable = true),
+    StructField("package_name",StringType, nullable = true),
+    StructField("parameters",ArrayType(StringType, containsNull = true), nullable = true)
+  ))
+
+  val minimumSQLTaskSchema: StructType = StructType(Seq(
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:354]: also add dashboards & alerts

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala [hunk 4]
```diff
@@ -258,11 +284,87 @@ object Schema extends SparkSessionWrapper {
...
+  ))
+
+  val minimumTaskDetailSchema: StructType = StructType(Seq(
+    StructField("notebook_task",minimumNotebookTaskSchema, nullable = true),
+    StructField("spark_python_task",minimumSparkPythonTaskSchema, nullable = true),
+    StructField("python_wheel_task",minimumPythonWheelTaskSchema, nullable = true),
+    StructField("spark_jar_task",minimumSparkJarTaskSchema, nullable = true),
+    StructField("spark_submit_task",minimumSparkSubmitTaskSchema, nullable = true),
+    StructField("shell_command_task",minimumShellCommandTaskSchema, nullable = true),
+    StructField("pipeline_task",minimumPipelineTaskSchema, nullable = true),
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:367]: add dbt & sql tasks

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala [hunk 5]
```diff
@@ -283,39 +385,119 @@ object Schema extends SparkSessionWrapper {
...
+    StructField("min_retry_interval_millis", LongType, nullable = true),
+    StructField("new_cluster", minimumNewClusterSchema, nullable = true),
+    StructField("notebook_task",minimumNotebookTaskSchema, nullable = true),
+    StructField("spark_python_task",minimumSparkPythonTaskSchema, nullable = true),
+    StructField("python_wheel_task",minimumPythonWheelTaskSchema, nullable = true),
+    StructField("spark_jar_task",minimumSparkJarTaskSchema, nullable = true),
+    StructField("spark_submit_task",minimumSparkSubmitTaskSchema, nullable = true),
+    StructField("shell_command_task",minimumShellCommandTaskSchema, nullable = true),
+    StructField("pipeline_task",minimumPipelineTaskSchema, nullable = true),
+    StructField("sql_task",minimumSQLTaskSchema, nullable = true)
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:404]: maybe also `dbt_task` ?

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala [hunk 6]
```diff
@@ -283,39 +385,119 @@ object Schema extends SparkSessionWrapper {
...
+    StructField("name", StringType, nullable = true),
+    StructField("new_cluster", minimumNewClusterSchema, nullable = true),
+    StructField("timeout_seconds", LongType, nullable = true),
+    StructField("notebook_task", minimumNotebookTaskSchema, nullable = true),
+    StructField("spark_python_task", minimumSparkPythonTaskSchema, nullable = true),
+    StructField("python_wheel_task", minimumPythonWheelTaskSchema, nullable = true),
+    StructField("spark_jar_task", minimumSparkJarTaskSchema, nullable = true),
+    StructField("spark_submit_task", minimumSparkSubmitTaskSchema, nullable = true),
+    StructField("shell_command_task", minimumShellCommandTaskSchema, nullable = true),
+    StructField("pipeline_task", minimumPipelineTaskSchema, nullable = true),
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:464]: add `sql_task` and `dbt_task`

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala [hunk 7]
```diff
@@ -283,39 +385,119 @@ object Schema extends SparkSessionWrapper {
...
+    StructField("new_cluster", minimumNewClusterSchema, nullable = true),
+    StructField("notebook_task", minimumNotebookTaskSchema, nullable = true),
+    StructField("schedule", minimumScheduleSchema, nullable = true),
+    StructField("notebook_task",minimumNotebookTaskSchema, nullable = true),
+    StructField("spark_python_task",minimumSparkPythonTaskSchema, nullable = true),
+    StructField("python_wheel_task", minimumPythonWheelTaskSchema, nullable = true),
+    StructField("spark_jar_task",minimumSparkJarTaskSchema, nullable = true),
+    StructField("spark_submit_task", minimumSparkSubmitTaskSchema, nullable = true),
+    StructField("shell_command_task",minimumShellCommandTaskSchema, nullable = true),
+    StructField("pipeline_task", minimumPipelineTaskSchema, nullable = true),
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:499]: potentially `dbt_task` and `sql_task` could be specified here as well

**Review by alexott — CHANGES_REQUESTED** [review]
first pass...

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala (hunk 6 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:464]: same as above, not selected anywhere; thus they don't need to be here -- if they become a downstream used filed, they will be added.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala (hunk 5 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:404]: minimumSchema

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala (hunk 2 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:305]: will add python_named_params but we're not fully support DLT yet and DLT and SQL tasks are still not really GA, I'm planning to give them some time to harden before I add them.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctions.scala (hunk 1 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctions.scala:180]: yeah that'll be broad to do when we add support for GCP -- we'll have to do that in several places.

Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctions.scala:180]: if you're talking about cloud_attributes -- it's been converted to a map due to high cardinality in some cases. This causes the schema to be ever larger -- we're already hitting limits on the old default of 4000 chars in schema so all key:value pairs with higher cardinality of String:String have intentionally been converted to map several versions ago.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala [hunk 8]
```diff
@@ -261,22 +262,30 @@ trait BronzeTransforms extends SparkSessionWrapper {
...
     val changeInventory = Map[String, Column](
-      "settings.new_cluster.custom_tags" -> SchemaTools.structToMap(outputDF, "settings.new_cluster.custom_tags"),
-      "settings.new_cluster.spark_conf" -> SchemaTools.structToMap(outputDF, "settings.new_cluster.spark_conf"),
-      "settings.new_cluster.spark_env_vars" -> SchemaTools.structToMap(outputDF, "settings.new_cluster.spark_env_vars"),
-      s"settings.new_cluster.aws_attributes" -> SchemaTools.structToMap(outputDF, s"settings.new_cluster.aws_attributes"),
-      s"settings.new_cluster.azure_attributes" -> SchemaTools.structToMap(outputDF, s"settings.new_cluster.azure_attributes"),
+      "settings.tasks" -> col("cleansedTasks"),
+      "settings.job_clusters" -> col("cleansedJobsClusters"),
+      "settings.tags" -> SchemaTools.structToMap(outputDF, "settings.tags"),
       "settings.notebook_task.base_parameters" -> SchemaTools.structToMap(outputDF, "settings.notebook_task.base_parameters")
```
[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala:277]: gotcha, thanks -- will do.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala (hunk 2 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:305]: also, jobs team is likely moving this to a new tag in the audit logs shortly, I'm working with them on the PR. Will likely change the way this works.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala (hunk 3 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:354]: minimum schema -- arrayType -- no risk of cardinality. When users have them, fields will appear.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala (hunk 4 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:367]: these may need to be handled in the conversion from struct to map but not necessary in the minSchema.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala (hunk 6 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Schema.scala:464]: actually -- this entire object should be deleted as it's no longer used. We converted golds to no longer reference the raw, `new_settings` field but instead hand pick the values we want. Also jobs team is changing the task details publication logic soon

File: src/main/scala/com/databricks/labs/overwatch/pipeline/WorkflowsTransforms.scala [hunk 9]
```diff
@@ -0,0 +1,1499 @@
...
+        'actionName,
+        'timestamp,
+        'jobId,
+        jobStatusBuildLookupLogic("jobName", lastJobStatus, rawDFSchema, Some("lookup_settings"), Some("name")),
+        jobStatusBuildLookupLogic("tags", lastJobStatus, rawDFSchema, Some("lookup_settings")),
+        fillForward("job_type", lastJobStatus),
+        fillForward("format", lastJobStatus, Seq(get_json_object('lookup_settings, "$.format"))),
+        jobStatusBuildLookupLogic("schedule", lastJobStatus, rawDFSchema, Some("lookup_settings")),
+        jobStatusBuildLookupLogic("email_notifications", lastJobStatus, rawDFSchema, Some("lookup_settings")),
+        'notebook_task,
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/WorkflowsTransforms.scala:270]: don't forget to add dbt & sql tasks later

File: src/main/scala/com/databricks/labs/overwatch/pipeline/WorkflowsTransforms.scala [hunk 10]
```diff
@@ -0,0 +1,1499 @@
...
+        $"settings.retry_on_timeout".alias("retry_on_timeout"),
+        $"settings.min_retry_interval_millis".alias("min_retry_interval_millis"),
+        struct(
+          $"settings.notebook_task".alias("notebook_task"),
+          $"settings.spark_python_task".alias("spark_python_task"),
+          $"settings.python_wheel_task".alias("python_wheel_task"),
+          $"settings.spark_jar_task".alias("spark_jar_task"),
+          $"settings.spark_submit_task".alias("spark_submit_task"),
+          $"settings.shell_command_task".alias("shell_command_task"),
+          $"settings.pipeline_task".alias("pipeline_task")
```
Comment (alexott) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/WorkflowsTransforms.scala:380]: same here for dbt & sql tasks

**Review by alexott — APPROVED** [review]
we need to fix tests although

**Pull request merged**